### PR TITLE
feat: support workload identity auth for Azure

### PR DESCRIPTION
**Pull request by wcy-fdu**
As title.

### CODE REVIEW

File: src/azure/storage/config.rs [hunk 1]
```diff
@@ -44,4 +44,22 @@ pub struct Config {
     ///
     /// This is part of use AAD(Azure Active Directory) authenticate on Azure VM
     pub endpoint: Option<String>,
+    /// `azure_federated_token` value will be loaded from:
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_FEDERATED_TOKEN`]
+    /// - profile config: `azure_federated_token_file`
+    pub azure_federated_token: Option<String>,
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:52]: Maybe we don't need `azure_` prefix?

File: src/azure/storage/config.rs [hunk 2]
```diff
@@ -44,4 +44,22 @@ pub struct Config {
...
+    /// `azure_federated_token_file` value will be loaded from:
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_FEDERATED_TOKEN_FILE`]
+    /// - profile config: `azure_federated_token_file`
+    pub azure_federated_token_file: Option<String>,
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:58]: We can just store `federated_token` inside.

File: src/azure/storage/config.rs [hunk 3]
```diff
@@ -44,4 +44,22 @@ pub struct Config {
...
+    /// `azure_tenant_id_env_key` value will be loaded from:
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_TENANT_ID_ENV_KEY`]
+    /// - profile config: `azure_tenant_id_env_key`
+    pub azure_tenant_id_env_key: Option<String>,
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:64]: This env is for `tenant_id`

File: src/azure/storage/workload_identity_credential.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    let endpoint = config.endpoint.as_deref().unwrap_or(MSI_ENDPOINT);
+
+    let mut query_items = vec![("api-version", MSI_API_VERSION), ("resource", resource)];
+    query_items.push(("token", &token));
+    query_items.push(("tenant_id", &tenant_id));
+    query_items.push(("client_id", &client_id));
+
+    let url = Url::parse_with_params(endpoint, &query_items)?;
+    let mut req = Request::builder()
```
Comment (Xuanwo) [inline src/azure/storage/workload_identity_credential.rs:50]: This seems incorrect. Where is `client_assertion` and `grant_type`?

File: src/azure/storage/config.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/azure/storage/config.rs:58]: What if user only set `federated_token_file` ? I saw in azure-sdk-rust, they check both `federated_token`  and `federated_token_file`, refer to https://github.com/Azure/azure-sdk-for-rust/blob/22ab1128f183fc80b97bb53b15486cc60702b61c/sdk/identity/src/token_credentials/workload_identity_credentials.rs#L92.

File: src/azure/storage/config.rs (hunk 3 above)
Comment (wcy-fdu) [reply src/azure/storage/config.rs:64]: Yes, is this required? Just saw that used in [WorkloadIdentityCredential](https://github.com/Azure/azure-sdk-for-rust/blob/22ab1128f183fc80b97bb53b15486cc60702b61c/sdk/identity/src/token_credentials/workload_identity_credentials.rs#L27) in azure sdk.

File: src/azure/storage/config.rs (hunk 2 above)
Comment (Xuanwo) [reply src/azure/storage/config.rs:58]: > What if user only set `federated_token_file` ? I saw in azure-sdk-rust, they check both `federated_token` and `federated_token_file`, refer to [Azure/azure-sdk-for-rust@`22ab112`/sdk/identity/src/token_credentials/workload_identity_credentials.rs#L92](https://github.com/Azure/azure-sdk-for-rust/blob/22ab1128f183fc80b97bb53b15486cc60702b61c/sdk/identity/src/token_credentials/workload_identity_credentials.rs#L92).

We can support loading from both `federated_token` and `federated_token_file` but only store it's content.

File: src/azure/storage/config.rs (hunk 3 above)
Comment (Xuanwo) [reply src/azure/storage/config.rs:64]: Not required, but for improve reading. Please use `tenant_id` and env `AZURE_TENANT_ID`

**Comment by wcy-fdu** [comment]
@Xuanwo  PTAL, thanks❤️

**Comment by Xuanwo** [comment]
> @Xuanwo PTAL, thanks❤️

Thanks, I will take a look tomorrow.

**Comment by wcy-fdu** [comment]
Can we apply this change to opendal side before this pr merged? Thus I can start follow-up testing as early as possible.

**Comment by Xuanwo** [comment]
> Can we apply this change to opendal side before this pr merged? Thus I can start follow-up testing as early as possible.

Yes, you can just pin reqsign to git version.

File: src/azure/storage/config.rs [hunk 5]
```diff
@@ -44,4 +47,56 @@ pub struct Config {
...
+    /// `tenant_id` value will be loaded from:
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_TENANT_ID`]
+    /// - profile config: `tenant_id`
+    pub tenant_id: Option<String>,
+    /// `authority_host` value will be loaded from:
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_AUTHORITY_HOST_ENV_KEY`]
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:65]: This should be `AZURE_AUTHORITY_HOST`

File: src/azure/storage/config.rs [hunk 6]
```diff
@@ -44,4 +47,56 @@ pub struct Config {
...
+    /// - profile config: `authority_host`
+    pub authority_host: Option<String>,
+}
+
+pub const AZURE_FEDERATED_TOKEN: &str = "AZURE_FEDERATED_TOKEN";
+pub const AZURE_FEDERATED_TOKEN_FILE: &str = "AZURE_FEDERATED_TOKEN_FILE";
+pub const AZURE_TENANT_ID: &str = "AZURE_TENANT_ID_ENV_KEY";
+pub const AZURE_AUTHORITY_HOST_ENV_KEY: &str = "AZURE_AUTHORITY_HOST_ENV_KEY";
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:73]: The same.

File: src/azure/storage/config.rs [hunk 7]
```diff
@@ -44,4 +47,56 @@ pub struct Config {
...
+    ///
+    /// - this field if it's `is_some`
+    /// - env value: [`AZURE_AUTHORITY_HOST_ENV_KEY`]
+    /// - profile config: `authority_host`
+    pub authority_host: Option<String>,
+}
+
+pub const AZURE_FEDERATED_TOKEN: &str = "AZURE_FEDERATED_TOKEN";
+pub const AZURE_FEDERATED_TOKEN_FILE: &str = "AZURE_FEDERATED_TOKEN_FILE";
+pub const AZURE_TENANT_ID: &str = "AZURE_TENANT_ID_ENV_KEY";
```
Comment (Xuanwo) [inline src/azure/storage/config.rs:72]: This should be `AZURE_TENANT_ID`

**Comment by Xuanwo** [comment]
Mostly LGTM, we need to setup the test.

**Comment by wcy-fdu** [comment]
Will reopen this pr after testing.

**Pull request closed without merging**